### ISSUE TITLE
Trying to reproduce structured3d finetuning results

### ISSUE THREAD

**Issue by chinmay0301ucsd** [manycore-research/SpatialLM#79]
Hi, I am trying to reduce the structured3d finetuning results and had some questions regarding it -- "

1.  Can you clarify if data augmentations were used during finetuning. 
2. If yes, does the cuboid crop augmentation matter a lot? Its not present in the released code hence asking. 

Thanks,
Chinmay

**Comment by chinmay0301ucsd** [comment]
So random_rotation is set to false right? That would result in using only
the discrete rotations.

**Comment by SamMaoYS** [comment]
yes

**Comment by chinmay0301ucsd** [comment]
Would it be possible to share the config file used for finetuning structured3d?

**Comment by SamMaoYS** [comment]
The training script we used differs from the released version, so the config file format may not be fully consistent. Here are some information that may help you reproduce the results.
For our experiments on Structured3D, we used a batch size of 8, trained for 50 epochs, with a learning rate of 1e-4 and a cosine learning rate scheduler (warm-up ratio 0.03), using AdamW as the optimizer. The number of bins during point cloud voxelization are reduced from 1280 to 640 for Structured3D finetuning, as the task doesn't require fine-grained object detection. You're welcome to experiment with different point cloud resolutions to evaluate performance trade-offs.

**Comment by chinmay0301ucsd** [comment]
I tried finetuning Spatiallm base model (1.1Qwen) on structured3d dataset with no augmentations (except scaling and discrete rotations), with lr 1e-4 for 50 iterations, with 640 bins, same optimizer, warmup and aforementioned batch settings.  

Layouts | F1@0.25 IoU | F1 @ 0.50 IoU
-------| ---- | ---------
Wall     | 0.89 | 0.8758
Door    | 0.6879 | 0.6643
Window| 0.602 | 0.5608
Average | 0.72694 | 0.70033

**Comment by SamMaoYS** [comment]
Hi, thank you for your effort in reproducing the results. I've uploaded the fine-tuned model to huggingface [ysmao/SpatialLM1.1-Qwen-0.5B-Structured3D-SFT](https://huggingface.co/ysmao/SpatialLM1.1-Qwen-0.5B-Structured3D-SFT), and the prediction results used for reporting the numbers here [structured3d-predictions](https://1drv.ms/u/c/d171cf5812b85e81/EdfZ8DaZSMRLnbpUT6PFVtMB4z5KABOt07XRgYclD2Culw?e=bXpKK6) for your reference. 

I plan to finetune another version using the released fine-tuning script. However, I'm currently quite busy and may not have the time in the following week or two. I will add this as a TODO item and will work on it when my schedule allows. Currently, I'm not sure what might be causing the performance gap between your results and ours. Hopefully, redoing the finetuning with the released script will help clarify that.

**Comment by SamMaoYS** [comment]
You can run the model using the following command:
`python inference.py -p path/to/test_point_clouds -o path/to/output_layouts --model_path SpatialLM1.1-Qwen-0.5B-Structured3D-SFT --inference_dtype float32`

**Comment by whu-pzhang** [comment]
> I tried finetuning Spatiallm base model (1.1Qwen) on structured3d dataset with no augmentations (except scaling and discrete rotations), with lr 1e-4 for 50 iterations, with 640 bins, same optimizer, warmup and aforementioned batch settings. My results have a gap compared to numbers in the paper, specially for doors and windows.
> 
> Layouts	F1@0.25 IoU	F1 @ 0.50 IoU
> Wall	0.89	0.8758
> Door	0.6879	0.6643
> Window	0.602	0.5608
> Average	0.72694	0.70033

I have done the same reproduction work as you, and my results are as follows. May I ask if you have now reached the desired scores?

| Layouts | F1 @.25 IoU        | F1 @.50 IoU        |
|---------|--------------------|--------------------|
| wall    | 0.8947179836727525 | 0.878849181126249  |
| door    | 0.6667609359920179 | 0.628435674379727  |
| window  | 0.5084176107089036 | 0.4605143285546706 |

**Comment by chinmay0301ucsd** [comment]
@whu-pzhang thanks for validating my results! No I haven't been able to match the numbers in the paper. I get similar numbers as you have reported.

**Comment by SamMaoYS** [comment]
Thank you for confirming the issue.
I will investigate this further. One possible cause I can think of is that: the pretraining was done on the SpatialLM dataset, including both single-room and multi-room data. The released checkpoint was finetuned with extra epochs on a subset of higher-quality single-room only data to better optimize for single-room settings. The Structured3D experiment used a checkpoint that was finetuned on the multi-room data with few extra epochs. Once the SpatialLM dataset is released, we will provide more detailed instructions on how to reproduce the experiments for both Structured3D and ScanNet experiments.

**Comment by chinmay0301ucsd** [comment]
Would it be possible to release the model checkpoint used in the paper for
obtaining structured3D results?

**Comment by SamMaoYS** [comment]
Hi @chinmay0301ucsd, @whu-pzhang, thanks for pointing out the issues. I also tried reproducing the results using the released training script and checkpoints, and noticed something unusual:

https://github.com/manycore-research/SpatialLM/blob/22a08ce9d12ef72e409f7d812790e004dd1691e7/spatiallm/layout/entity.py#L70

In this line, when scaling the wall entity, I forgot to apply the scaling to by and bz by mistake. This might be the cause of the problem. I’ll fix this bug and re-run the experiments on my side to confirm.

I am sincerely sorry for the silly bug, and waste your valuable time, and thanks again for catching this.

**Comment by ZhouBay-TF** [comment]
> > 我尝试在 Structured3D 数据集上对 Spatiallm 基础模型 (1.1Qwen) 进行微调，没有进行任何数据增强（缩放和离散旋转除外），使用 LR 1e-4 进行 50 次迭代，使用 640 个 bin，使用相同的优化器、预热和前面提到的批处理设置。我的结果与论文中的数字相比存在差距，尤其是在门窗方面。
> > 布局 F1@0.25 IoU F1 @ 0.50 IoU
> > 墙 0.89 0.8758
> > 门 0.6879 0.6643
> > 窗 0.602 0.5608
> > 平均值 0.72694 0.70033
> 
> 我做了和你一样的复现工作，结果如下。请问你现在达到预期分数了吗？
> 
> 布局	F1 @.25 IoU	F1 @.50 IoU
> 墙	0.8947179836727525	0.878849181126249
> 门	0.6667609359920179	0.628435674379727
> 窗户	0.5084176107089036	0.4605143285546706

hello, how much video GPU does this training and reasoning require? Thank you

**Comment by SamMaoYS** [comment]
I also redo the experiment with the released training script and starting from the checkpoint above. Here are the numbers I got:
| Layouts | F1 @.25 IoU | F1 @.50 IoU |
|---------|-------------|-------------|
| wall | 0.954239374314068 | 0.9437889645930915 |
| door | 0.8156503638457702 | 0.790702494406211 |
| window | 0.7418540979304805 | 0.7121263638039519 |

The window and door scores are slightly lower than those reported in the paper by a few percentage points. I’m not sure whether this is due to variance across training runs or some minor inconsistencies in the training script. I’ll look into this further when I have more time. If you’re interested in reproducing the results, please let me know how it goes, and feel free to reach out with any further requests or questions.

**Comment by yeeeeeeii** [comment]
> Hi, thank you for your effort in reproducing the results. I've uploaded the fine-tuned model to huggingface [ysmao/SpatialLM1.1-Qwen-0.5B-Structured3D-SFT](https://huggingface.co/ysmao/SpatialLM1.1-Qwen-0.5B-Structured3D-SFT), and the prediction results used for reporting the numbers here [structured3d-predictions](https://1drv.ms/u/c/d171cf5812b85e81/EdfZ8DaZSMRLnbpUT6PFVtMB4z5KABOt07XRgYclD2Culw?e=bXpKK6) for your reference.
> 
> I plan to finetune another version using the released fine-tuning script. However, I'm currently quite busy and may not have the time in the following week or two. I will add this as a TODO item and will work on it when my schedule allows. Currently, I'm not sure what might be causing the performance gap between your results and ours. Hopefully, redoing the finetuning with the released script will help clarify that.

Hi,I ran eval.py using [structured3d-predictions](https://1drv.ms/u/c/d171cf5812b85e81/EdfZ8DaZSMRLnbpUT6PFVtMB4z5KABOt07XRgYclD2Culw?e=bXpKK6) and obtained outcomes are as follows. However, these results seem higher than those reported in the paper. Is this normal? 
| Layouts | F1 @.25 IoU        | F1 @.50 IoU        |
|-------|-------|-------|
| wall    | 0.9527221819169637 | 0.9452651637911722 |
| door    | 0.9585782543993591 | 0.9565706262506521 |
| window  | 0.9173996506274097 | 0.9023430905587376 |

**Comment by SamMaoYS** [comment]
@chinmay0301ucsd helped and identified a bug in the original evaluation script. After fixing it, doors and windows will have higher scores than those reported in the initial report. We apologize for the confusion caused by the inconsistent numbers and will likely update the report in the future.

**Comment by anac0der** [comment]
@SamMaoYS Are the updated metrics for your model available now, considering the changes in the script? I'm especially interested in the experiment with training on Structured3D from scratch

**Comment by SamMaoYS** [comment]
Hi @anac0der, thank you for your interest. Here is a preview of the updated results:

<img width="693" height="253" alt="Image" src="https://github.com/user-attachments/assets/6cedcbad-acf6-4399-9ac3-f20164131f67" />

**Comment by 920902021** [comment]
@SamMaoYS Hi, it seems like the link for the results has expired. Could you please provide a new one? Specifically for the "structured3d-predictions". Thanks a lot!

**Comment by SamMaoYS** [comment]
@920902021 Sure, here is the file. [structured3d-predictions.zip](https://1drv.ms/u/c/d171cf5812b85e81/IQDX2fA2mUjES526VE-jxVbTAcTIk2OcdMWV08BQECP8diU?e=tocW4l)

**Comment by Dominic23331** [comment]
> Hi [@anac0der](https://github.com/anac0der), thank you for your interest. Here is a preview of the updated results:
> 
> <img alt="Image" width="693" height="253" src="https://private-user-images.githubusercontent.com/36813656/492122328-6cedcbad-acf6-4399-9ac3-f20164131f67.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Nzk0NDQxNTAsIm5iZiI6MTc3OTQ0Mzg1MCwicGF0aCI6Ii8zNjgxMzY1Ni80OTIxMjIzMjgtNmNlZGNiYWQtYWNmNi00Mzk5LTlhYzMtZjIwMTY0MTMxZjY3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTIyVDA5NTczMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1ODdmMGRlNDQxNTZiNGE3NzJkM2FmMDhmMzVhNTI4MDAwY2I4OGUyZjE5MjNmMDczNTk2ODAxZjI4NDg1OGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.0CTX8U4vvs6Pjk24J_3OASo6MUA-c-bhId7z8n_9NDY">

请问在structure3d数据集中训练的模型权重有计划公开吗？

**Comment by tmzcometolearn** [comment]
> I also redo the experiment with the released training script and starting from the checkpoint above. Here are the numbers I got:
> 
> Layouts	F1 @.25 IoU	F1 @.50 IoU
> wall	0.954239374314068	0.9437889645930915
> door	0.8156503638457702	0.790702494406211
> window	0.7418540979304805	0.7121263638039519
> The window and door scores are slightly lower than those reported in the paper by a few percentage points. I’m not sure whether this is due to variance across training runs or some minor inconsistencies in the training script. I’ll look into this further when I have more time. If you’re interested in reproducing the results, please let me know how it goes, and feel free to reach out with any further requests or questions.

请问这个结果，是在https://huggingface.co/ysmao/SpatialLM1.1-Qwen-0.5B-Layout-Intmd 的基础上，用https://huggingface.co/datasets/ysmao/structured3d-spatiallm 数据集，进行了50个epoch的微调得到的结果吗？（也就是这个ckpt：https://huggingface.co/ysmao/SpatialLM1.1-Qwen-0.5B-Structured3D-SFT ）。谢谢~

**Comment by SamMaoYS** [comment]
是的
